tests/smart_components_test.py

```diff
@@ -0,0 +1,179 @@
+    wide.paths.append(rectangle_path(100, 100, 500, 100))
+    wide.smartComponentPoleMapping["Width"] = 2
+    wide.smartComponentPoleMapping["Height"] = 1
+    wide.smartComponentPoleMapping["Shift"] = 2
+
+    tall = GSLayer()
+    tall.name = "Tall"
+    tall.layerId = "tall"
+    tall.associatedMasterId = font.masters[0].id
+    tall.width = 300
+    rectangle.layers.append(tall)
+    tall.paths.append(rectangle_path(100, 100, 100, 500))
+    tall.smartComponentPoleMapping["Width"] = 1
+    tall.smartComponentPoleMapping["Height"] = 2
+    tall.smartComponentPoleMapping["Shift"] = 2
+
+    shifted = GSLayer()
+    shifted.name = "Shifted"
+    shifted.layerId = "shifted"
+    shifted.associatedMasterId = font.masters[0].id
+    shifted.width = 100
+    rectangle.layers.append(shifted)
+    shifted.paths.append(rectangle_path(0, 0, 100, 100))
+    shifted.smartComponentPoleMapping["Width"] = 1
+    shifted.smartComponentPoleMapping["Height"] = 1
+    shifted.smartComponentPoleMapping["Shift"] = 1
+
+    # Also add a normal glyph in which we can instanciate the component
+    a = GSGlyph()
+    a.name = "a"
+    font.glyphs.append(a)
+    regular = GSLayer()
+    regular.layerId = font.masters[0].id
+    regular.associatedMasterId = font.masters[0].id
+    regular.width = 1000
+    a.layers.append(regular)
+
+    component = GSComponent(rectangle.name)
+    component.smartComponentValues = {}
+    regular.components.append(component)
+
+    return font
+
+
+def rectangle_path(x, y, w, h):
+    path = GSPath()
+    path.nodes.append(GSNode((x, y)))
+    path.nodes.append(GSNode((x + w, y)))
+    path.nodes.append(GSNode((x + w, y + h)))
+    path.nodes.append(GSNode((x, y + h)))
+    return path
+
+
+def test_dump_smart_font(smart_font, tmpdir):
+    smart_font.save("smart.glyphs")
+
+
+@pytest.mark.parametrize(
+    "values,expected",
+    [
+        # Eight corners
+        ({"Width": 0, "Height": 100, "Shift": 0}, (100, 100, 100, 100)),
+        ({"Width": 1, "Height": 100, "Shift": 0}, (100, 100, 500, 100)),
+        ({"Width": 0, "Height": 500, "Shift": 0}, (100, 100, 100, 500)),
+        ({"Width": 1, "Height": 500, "Shift": 0}, (100, 100, 500, 500)),
+        ({"Width": 0, "Height": 100, "Shift": -100}, (0, 0, 100, 100)),
+        ({"Width": 1, "Height": 100, "Shift": -100}, (0, 0, 500, 100)),
+        ({"Width": 0, "Height": 500, "Shift": -100}, (0, 0, 100, 500)),
+        ({"Width": 1, "Height": 500, "Shift": -100}, (0, 0, 500, 500)),
+        # Some points in the middle
+        ({"Width": 0.5, "Height": 300, "Shift": -50}, (50, 50, 300, 300)),
+    ],
+)
+def test_smart_component_regular(values, expected, smart_font):
+    component = smart_font.glyphs["a"].layers[0].components[0]
+    for key, value in values.items():
+        component.smartComponentValues[key] = value
+
+    (ufo,) = to_ufos(smart_font)
+
+    assert get_rectangle_data(ufo) == expected
+
+
+def get_rectangle_data(ufo, glyph_name="a", component_index=0):
```

Comment:
`component_index` parameter is unused in the body of the `get_rectangle_data` - remove it?